presto-main/src/main/java/com/facebook/presto/operator/unnest/ArrayUnnester.java

```diff
@@ -46,33 +50,51 @@ public int getChannelCount()
     }
 
     @Override
-    public long getRetainedSizeInBytes()
+    public void resetInput(Block block)
     {
-        // The lengths array in unnestBlockBuilders is the same object as in the unnester and doesn't need to be counted again.
-        return INSTANCE_SIZE + sizeOf(getLengths());
+        requireNonNull(block, "block is null");
+
+        columnarArray = toColumnarArray(block);
+        unnestBlockBuilder.resetInputBlock(columnarArray.getElementsBlock());
     }
 
     @Override
-    protected void resetColumnarStructure(Block block)
+    public int[] populateLengths(int positionCount)
```

Comment:
I find this method awkward. Looking at how it is used, I'm seeing that it is always called after resetInput. Hence, I suggest to fold this logic into resetInput. I also see that UnnestBlockBuilder uses lengths array only in buildOutputBlockWithNulls method. Hence, I suggest to remove setLengths methods and pass lengths to buildOutputBlockWithNulls method directly.

The interface would looks like this:

```
void resetInput(Block block, int positionCount);

int[] getLengths();
```

and the implementation:

```
    @Override
    public void resetInput(Block block, int positionCount)
    {
        requireNonNull(block, "block is null");

        columnarArray = toColumnarArray(block);
        unnestBlockBuilder.resetInputBlock(columnarArray.getElementsBlock());

        lengths = ensureCapacity(lengths, positionCount);
        for (int i = 0; i < positionCount; i++) {
            lengths[i] = columnarArray.getLength(i);
        }
    }

    @Override
    public int[] getLengths()
    {
        return lengths;
    }

    @Override
    public Block[] buildOutputBlocks(int[] maxEntries, int startPosition, int batchSize, int currentBatchTotalEntries)
    {
        boolean nullRequired = columnarArray.getOffset(startPosition + batchSize) - columnarArray.getOffset(startPosition) < currentBatchTotalEntries;

        Block[] outputBlocks = new Block[1];
        if (nullRequired) {
            outputBlocks[0] = unnestBlockBuilder.buildOutputBlockWithNulls(maxEntries, startPosition, batchSize, currentBatchTotalEntries, lengths);
        }
        else {
            outputBlocks[0] = unnestBlockBuilder.buildOutputBlockWithoutNulls(currentBatchTotalEntries);
        }

        return outputBlocks;
    }
```